### PR TITLE
Cert Refresh

### PR DESCRIPTION
**Pull request by lindig**
Just for discussion. This is my approach to refresh server certificates. This rotation of the certs works both on the client and the server, including updating bundles and the database. However, I don't know yet how best to activate them without losing connections. Hence, the stunnel re-configuration is commented out.

### CODE REVIEW

File: ocaml/xapi/xapi_host.ml [hunk 1]
```diff
@@ -1465,7 +1465,11 @@ let replace_host_certificate ~__context ~type' ~host
   let old_certs = Db_util.get_host_certs ~__context ~type' ~host in
   let new_cert = write_cert_fs () in
```
Comment (psafont) [inline ocaml/xapi/xapi_host.ml:1466]: This overwrites the existing certificate. What happens when there's a failure and the previous certificates needs to be reinstated?

File: ocaml/idl/datamodel_host.ml [hunk 2]
```diff
@@ -1029,6 +1029,14 @@ let host_query_ha = call ~flags:[`Session]
       ~allowed_roles:_R_READ_ONLY
       ()
 
+  let refresh_server_certificates = call
+      ~lifecycle:[Published, rel_next, ""]
+      ~name:"refresh_server_certificates"
+      ~doc:"Replace the self-signed certficates for the host with new ones."
+      ~params:[Ref _host, "host", "The host"]
+      ~allowed_roles:_R_POOL_ADMIN
```
Comment (psafont) [inline ocaml/idl/datamodel_host.ml:1037]: This should not be documented, IMO. Users will find confusing having "refresh_server_certificates" and "reset_server_certificate".
I think recycling certificates in this way only makes sense for internal certificates, dropping the 's' among other changes would make it less confusing even for us.

File: ocaml/xapi/xapi_host.ml (hunk 1 above)
Comment (lindig) [reply ocaml/xapi/xapi_host.ml:1466]: Maybe I should rename them to `*.bak` such that they would not be picked up by bundling but they would be around for recovery.

File: ocaml/idl/datamodel_host.ml (hunk 2 above)
Comment (lindig) [reply ocaml/idl/datamodel_host.ml:1037]: Ok - will limit this to the internal certificates. But to be clear: a server has two self-signed certificates, one for internal and one for external use. So don't rotate the external one? I don't think this makes sense from a security perspective.

Comment (psafont) [reply ocaml/idl/datamodel_host.ml:1037]: The external one is already handled by the call `reset_server_certificate`. It's a different call because users might have installed their own certificates and they just want to replace the self-signed ones without affecting the external certificates.

File: ocaml/xapi/message_forwarding.ml [hunk 3]
```diff
@@ -3546,6 +3546,25 @@ functor
             Client.Host.get_server_certificate rpc session_id host
         )
 
+      let refresh_server_certificate ~__context ~host =
```
Comment (lippirk) [inline ocaml/xapi/message_forwarding.ml:3549]: We probably want a lock to prevent concurrent refreshes

Comment (lippirk) [reply ocaml/xapi/message_forwarding.ml:3549]: Maybe we want to error if a refresh is already in progress

File: ocaml/xapi/cert_distrib.ml [hunk 4]
```diff
@@ -439,3 +439,17 @@ let import_joining_pool_ca_certificates ~__context ~ca_certs =
...
   Worker.local_regen_bundle ~__context
+
+let distribute_new_host_cert ~__context ~host ~content =
+  let hosts = Db.Host.get_all ~__context in
+  let uuid = Db.Host.get_uuid ~__context ~self:host in
+  let file =
+    WireProtocol.{filename= Printf.sprintf "%s.new.pem" uuid; content}
+  in
+  let job rpc session_id host =
+    Worker.remote_write_certs_fs HostPoolCertificate Merge [file] host rpc
```
Comment (lippirk) [inline ocaml/xapi/cert_distrib.ml:450]: what happens if `file` already exists on the remote host? We probably don't want to overwrite it, because the remote host could be relying on it

File: ocaml/xapi/cert_refresh.ml [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  let ips = [ip] in
+  Gencertlib.Selfcert.host ~name ~dns_names ~ips path
+
+let host ~__context ~type' =
+  let host = Helpers.get_localhost ~__context in
+  let dbg = Context.string_of_task __context in
+  let pem = cert_path type' in
+  let path = new_cert_path type' in
+  let cert = new_host_cert ~dbg ~path in
+  let bak = backup_cert_path type' in
```
Comment (lippirk) [inline ocaml/xapi/cert_refresh.ml:59]: We might want to think about the case where `bak` exists on the file system at this point - it probably means that a previous cert refresh failed. Do we just error out in this case or try to resolve the problem?

Comment (lippirk) [reply ocaml/xapi/cert_refresh.ml:59]: Ah I see we don't actually remove `bak` at the end. The idea is that it is useful to keep around in case the user needs to manually intervene because of a failure?

I'm thinking about the case where a cert refresh has failed - the user's first instinct after running `xe cert-refresh` and seeing a failure is going to be to run it again, so it would be nice not to overwrite `bak` in this case

Comment (lindig) [reply ocaml/xapi/cert_refresh.ml:59]: Good point! Maybe we should fail if the backup exists

File: ocaml/xapi/cert_refresh.ml [hunk 6]
```diff
@@ -50,6 +50,12 @@ let new_host_cert ~dbg ~path : X509.Certificate.t =
   let ips = [ip] in
   Gencertlib.Selfcert.host ~name ~dns_names ~ips path
 
+let reload_stunnel () =
+  let systemctl cmd =
+    Helpers.call_script !Xapi_globs.systemctl [cmd; "stunnel@xapi"] |> ignore
+  in
+  systemctl "reload-or-restart"
+
```
Comment (lippirk) [inline ocaml/xapi/cert_refresh.ml:58]: For consistency you could put this function in `Helpers.Stunnel`

We'll probably want to use it here too https://github.com/xapi-project/xen-api/blob/99d1f2d9c34265ebf0f78650960f3a4b5fb8fc87/ocaml/xapi/cert_distrib.ml#L229 (though that is out of scope for this PR)

File: ocaml/xapi/cert_distrib.mli [hunk 7]
```diff
@@ -50,3 +50,7 @@ val import_joining_pool_ca_certificates :
     This parameter must be the result of
     [exchange_ca_certificates_with_joiner]. This function was designed
     as part of pool join and is unlikely to be useful elsewhere. *)
+
+val distribute_new_host_cert :
+  __context:Context.t -> host:[`host] API.Ref.t -> content:string -> unit
+(** distribute a new (additioanl) certificate for [host] in the pool *)
```
Comment (lippirk) [inline ocaml/xapi/cert_distrib.mli:56]:
```suggestion
(** distribute a new (additional) certificate for [host] in the pool *)
```

Comment (lindig) [reply ocaml/xapi/cert_distrib.mli]: Fixed it.

File: ocaml/xapi/cert_refresh.ml (hunk 6 above)
Comment (lindig) [reply ocaml/xapi/cert_refresh.ml]: Done

**Comment by lindig** [comment]
This should not be merged before we have support for systemd-based reload of Stunnel merged.

File: ocaml/xapi/message_forwarding.ml (hunk 3 above)
Comment (lippirk) [reply ocaml/xapi/message_forwarding.ml:3549]: your call to `with_pool_operation` should error if a refresh is already in progress

**Review by lippirk — APPROVED** [review]
We should use the cert distrib lock introduced here: https://github.com/xapi-project/xen-api/pull/4441/files . Whilst it seems like the lock induced by the cert refresh pool operation should be sufficient, it does not mutually exclude pool.join cert distribs and Host.refresh_server_certificate

However I'm happy to get this in and fix that later

**Review by psafont — APPROVED** [review]
